core/src/main/java/com/netflix/iceberg/BaseTransaction.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.netflix.iceberg.exceptions.CommitFailedException;
+import com.netflix.iceberg.io.InputFile;
+import com.netflix.iceberg.io.OutputFile;
+import com.netflix.iceberg.util.Tasks;
+import java.util.List;
+import java.util.Map;
+
+import static com.netflix.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS;
+import static com.netflix.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS_DEFAULT;
+import static com.netflix.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS;
+import static com.netflix.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS_DEFAULT;
+import static com.netflix.iceberg.TableProperties.COMMIT_NUM_RETRIES;
+import static com.netflix.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static com.netflix.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static com.netflix.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+// TODO: ensure this correctly cleans up unused manifests created by SnapshotUpdates
+class BaseTransaction implements Transaction {
+  static Transaction createTableTransaction(TableOperations ops, TableMetadata start) {
+    Preconditions.checkArgument(ops.current() == null,
+        "Cannot start create table transaction: table already exists");
+    return new BaseTransaction(ops, start);
+  }
+
+  static Transaction newTransaction(TableOperations ops) {
+    return new BaseTransaction(ops, ops.refresh());
+  }
+
+  private final TransactionTable transactionTable;
+  private final TableOperations ops;
+  private final TableOperations transactionOps;
+  private final List<PendingUpdate> updates;
+  private TableMetadata base;
+  private TableMetadata lastBase;
+  private TableMetadata current;
+
+  private BaseTransaction(TableOperations ops, TableMetadata start) {
+    this.transactionTable = new TransactionTable();
+    this.ops = ops;
+    this.transactionOps = new TransactionTableOperations();
+    this.updates = Lists.newArrayList();
+    this.base = ops.current();
+    this.lastBase = null;
+    this.current = start;
+  }
+
+  @Override
+  public Table table() {
+    return transactionTable;
+  }
+
+  private void checkLastOperationCommitted(String operation) {
+    Preconditions.checkState(lastBase != current,
```

Comment:
So we only want users of this method to 
`beginTransction();
updateProperties();
updateProperties();
commit(); `

and we do not want users to do 

`
beginTransaction()
updateProperties()
append()
commit()
`
From the API documentation it is not very clear that the semantics require users to commit a transaction with one type of operation. Also other than simplicity any reason we should not support the second case?

Reply:
The case that we're avoiding is slightly more complicated. A transaction is made up of a series of pending updates. We want to make sure it is clear that those updates are each a separate and atomic action. This ensures that you `commit` one action before preparing another so it is clear the order in which those actions will be applied to the table.

Otherwise, I could:
1. Create an append, add some files
2. Create a delete, delete some files, commit the delete
3. Add more files to the append, and commit the append

That raises questions like: which action is applied first, the append (created first) or the delete (completed first)? Rather than worry about these, you have to create and complete each update in the same order. To do that, we ensure that you've updated the state by committing before we will create another update.